src/fm.py

```diff
@@ -58,6 +58,30 @@ def sa_construct(x: str) -> list[int]:
     return [len(x) - len(i) for i in suf]
 
 
+def fm_search(x: rotating_string, p: str, sa: list[int]) -> list[int]:
+    # make c and o table (this should be moved to pre processing)
+    alpha = sorted(set(x.x))
+    c = {a: 0 for a in alpha}
+    for a in x.x:
+        c[a] += 1
+    accsum = 0
+    for bucket in c:
+        c[bucket], accsum = accsum, accsum + c[bucket]
+
+    # should probably stop using dict for this, and just map them down to indices in a list...
+    O = [{a: 0 for a in c} for _ in range(len(x)+1)]
+    for i, o in enumerate(O):
+        if i != 0 and i < len(x):
+            o[x[sa[i]]] = O[i-1][x[sa[i]]] + (x[sa[i]] == x[sa[i-1]])
```

Comment:
Does this work? I see both a small `o` and a big `O`, but I don't see you creating any `o`.